src/Tests/Microsoft.Diagnostics.Monitoring.UnitTests/HttpApi/ResponseStreamHolder.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.UnitTests.HttpApi
+{
+    /// <summary>
+    /// Holds a <see cref="System.IO.Stream"/> from a <see cref="HttpResponseMessage"/> to
+    /// ensure that the message and stream are disposed together.
+    /// </summary>
+    internal class ResponseStreamHolder : IDisposable
+    {
+        private readonly HttpResponseMessage _response;
+
+        public Stream Stream { get; private set; }
+
+        private ResponseStreamHolder(HttpResponseMessage response)
+        {
+            _response = response ?? throw new ArgumentNullException(nameof(response));
+        }
+
+        public void Dispose()
+        {
+            // The response disposes the stream when disposed.
+            _response.Dispose();
+        }
+
+        public static async Task<ResponseStreamHolder> CreateAsync(DisposableBox<HttpResponseMessage> responseBox)
+        {
+            using DisposableBox<ResponseStreamHolder> holderBox = new(new(responseBox.Release()));
```

Comment:
I'm not sure why we need holder box here? Doesn't the lifetime ultimately transfer to ResponseStreamHolder?

Reply:
It does, but if `holderBox.Value._response.Content.ReadAsStreamAsync().ConfigureAwait(false);` throws an exception, we want to dispose the ResponseStreamHolder (since it owns the HttpResponseMessage at that point. If it doesn't throw, then we want to not dispose ResponseStreamHolder and just return it.